### ISSUE TITLE
"@OA\\Items() parent type must be "array" in "

### ISSUE THREAD

**Issue by Aerendir** [api-platform/core#6194]
**API Platform version(s) affected**: 3.2.* (still present in 3.2.15)

**Description**  

```console
Caused by
 ErrorException: @OA\Items() parent type must be "array" in 
```

**How to reproduce**  

Create an entity that uses an embeddable.

```php
<?php

declare(strict_types=1);

namespace App\Entity;

use ApiPlatform\Metadata as API;
use App\Entity\Embeddable\TestEmbeddable;

#[API\ApiResource]
class TestResource
{
    private ?TestEmbeddable $testEmbeddable = null;

    public function setTestEmbeddable(?TestEmbeddable $testEmbeddable): self
    {
        $this->testEmbeddable = $testEmbeddable;

        return $this;
    }
}
```

```php
<?php

declare(strict_types=1);

namespace App\Entity\Embeddable;

class TestEmbeddable
{
    private ?array $testArrayOrNull = null;

    public function getTestArrayOrNull(): ?array
    {
        return $this->testArrayOrNull;
    }

    /**
     * Here, removing the "?" (nullability) makes the error disappear
     */
    private function setTestArrayOrNull(?array $arrayOrNull): void
    {
        $this->testArrayOrNull = $arrayOrNull;
    }
}
```

Trying to get the documentation in Swagger, causes the error.

The error is triggered in `zircote/swagger-php`:

https://github.com/zircote/swagger-php/blob/a6d8a93e5daa83d3b959221b0f8f176953d145e9/src/Annotations/Items.php#L53-L56

It expects `$parent->type` be `array` while it actually is an `array` with two values:

![Screenshot 2024-03-01 alle 17 53 59](https://github.com/api-platform/core/assets/1968622/97461a9d-a6fe-401c-948e-8621f0832216)

The fact that `TestEmbeddable::setTestArrayOrNull()` accepts an `array` or a `null` value causes `$parent->type` to have an array with the two possible acceptable values: `array` and `null`.

**With `api-platform/core:3.1.*` this error didn't happened.**

**Possible Solution**  
I think the problem may be in `ApiPlatform\JsonSchema\SchemaFactory`, but I'm not able to understand what changed that now causes the error.

**Additional Context**  

The code that adds the nullability is this:

https://github.com/api-platform/core/blob/bfd759b627990926fddd11c6737af8f993af7bb5/src/JsonSchema/Metadata/Property/Factory/SchemaPropertyMetadataFactory.php#L278-L282

Is here that `$type->type` is not `=== 'array'` but is equal to `['array', 'null']` and this breaks `zircote/swagger-php`.

I don't know if the issue has to be fixed in `zircote/swagger-php` or here, in `api-platform/core`.

Digging deeper in this issue, I noticed the add of `null` is done on purpose (correctly):

https://github.com/api-platform/core/commit/d793ffb9228a21655ee35f0b90a959f93281a4cf#diff-32d96682e97b6e04a17576ec737dd6cd19face8782e30713fe1fbf254e9ffaf2R89

https://github.com/api-platform/core/commit/d793ffb9228a21655ee35f0b90a959f93281a4cf#diff-32d96682e97b6e04a17576ec737dd6cd19face8782e30713fe1fbf254e9ffaf2R241

## The root cause of the issue

> # Swap nullable for type arrays
> In line with JSON Schema, the `type` keyword can now define multiple types for a schema with an array. This is useful new functionality, but has also made nullable redundant. In order to achieve the goal of letting JSON Schema tools understand OpenAPI, it was decided to remove nullable entirely instead of deprecate it. 
> 
> ```
> # OpenAPI v3.0
> type: string
> nullable: true
> ```
> 
> ```
> # OpenAPI v3.1
> type:
> - "string"
> - "null" 
> ```
> 
> This follows the keyword independence concept, where keywords should only add constraints, but adding one more keyword should not remove a constraint. Find and replace should solve this one rather quickly.
>
> *Source: https://www.openapis.org/blog/2021/02/16/migrating-from-openapi-3-0-to-3-1-0*

I'm not sure how to deal with this, but it seems to me that ApiPlatform generates OpenApi docs for version 3.1, while `zircote/swagger-php` doesn't still fully support it.

Unsure where (and what) to fix this 🤔

## Some usefult links

- https://github.com/zircote/swagger-php/issues/885
- https://github.com/zircote/swagger-php/issues/1257
- https://github.com/zircote/swagger-php/issues/1354
- https://github.com/api-platform/core/pull/5470
- https://github.com/api-platform/core/pull/5977

## (Possible) Solution

Use the Swagger UI integrated in ApiPlatform:

- https://api-platform.com/docs/core/openapi/

The types returned are not precise, but, at least, it works...

**Comment by soyuka** [comment]
`type` can be an array if we follow the json schema specification, I don't think I can do much here.

**Comment by Aerendir** [comment]
Yes, I saw the version 3.1 of the OpenAPI specification: it now allows an array.
The problem is that they changed the type of the field, but this is a breaking change as `swagger-php` is now broken.

When `api-platform` used the version 3.0, it was possible to use also `nelmio/api-docs-bundle` (that, in turn, use `swagger-php`).

But, as `api-platform` upgraded to version 3.1 of the OpenAPI specification, passing an array with the two possible types, `swagger-php` broke.

I see only two possible paths:

- `api-platform` should allow to choose the version of the OpenAPI specification (`3.0|3.1`): this way, the retro-compatibility is maintained
- `swagger-php` has to upgrade the code to support also the version `3.1` (there is an open task since the 2020 about this support)

For the future, for sure `api-platform` should take into account that also a minor version may break existing documentations.

The biggest problem, however, is that it is not possible to document external endpoints with `api-platform` (not entirely, at least): this is another are where `api-platform` should improve.

One problem, many causes as of what it seems.

In the meantime, I dropped `nelmio/api-docs-bundle` and gone with SwaggerUI integrated in `api-platform`, but now I don't have a full nor precise documentation for custom operations (that, unfortunately, cannot be addressed in a clean way using custom processors).

**Comment by dreamworks** [comment]
I see no sense in autoclosing this "issue"; it should be at least be visible .. 
That's my main reason why I add a note :)

**Comment by Aerendir** [comment]
This is issue should be managed, I agree.

**Comment by Aerendir** [comment]
Keep open

**Comment by Francexi** [comment]
By the way, this seems to solved if you do something like this

```diff
1c1
<         if ($parent instanceof Schema && $parent->type !== 'array') {
---
>         if ($parent instanceof Schema && ($parent->type !== 'array' && (is_array($parent->type) && !empty(array_diff($parent->type, ['array', "null"]))))) {
5d4
<
```

in here

https://github.com/zircote/swagger-php/blob/master/src/Annotations/Items.php#L53

is quick and dirty, I didn't check for version or anything

**Comment by soyuka** [comment]
Closing — partial fix landed in #8225 (merged into 4.3), shipping in 4.3.8.

## What was fixed

API Platform's OpenAPI 3.0 downgrade path (`?spec_version=3.0.0` or `api:openapi:export --spec-version=3.0.0`) used to emit invalid 3.0 schemas: orphan `items`, leaked `type: 'null'` entries (not valid in 3.0), no recursion into nested embeddables / `items` / `allOf`/`oneOf`/`anyOf`, no walk of path-level inline schemas.

`LegacyOpenApiNormalizer` is now a recursive walk that maps nullable types to `type + nullable: true` (idiomatic 3.0), preserves `items`/`properties`/`additionalProperties` on nullable arrays and objects, recurses through all schema-bearing keys, and walks `requestBody`/`responses`/`parameters` schemas. The embeddable case from this issue (`?array $testArrayOrNull`) now produces:

```yaml
testArrayOrNull:
  type: array
  items: { ... }
  nullable: true
```

## Why this doesn't fully close the upstream concern

The root error (`@OA\Items() parent type must be "array"`) originates in `zircote/swagger-php`'s `Items` annotation, which still expects scalar `type === 'array'` and doesn't understand 3.1's `type: ['array', 'null']`. That's a `nelmio/api-doc-bundle` + `swagger-php` integration problem — they consume PHP attributes directly via `swagger-php`, not API Platform's JSON output, so our downgrade can't reach them.

Tracked upstream: zircote/swagger-php#885, #1257, #1354.

## Workarounds

- Consumers that read API Platform's JSON output (Azure APIM, generators, etc.): request `?spec_version=3.0.0` — now produces spec-valid 3.0.
- `nelmio/api-doc-bundle` integration: pending upstream `swagger-php` 3.1 support. Built-in SwaggerUI is the supported path in the meantime (https://api-platform.com/docs/core/openapi/).
